System/Process/CommunicationHandle.hsc

```diff
@@ -0,0 +1,371 @@
+##if !defined(mingw32_HOST_OS)
+  (ourHandle, theirHandle) <- mbSwap <$> createPipe
+  -- Don't allow the child process to inherit a parent file descriptor
+  -- (such inheritance happens by default on Unix).
+  ourFD   <- Fd . fdFD <$> handleToFd ourHandle
+  setFdOption ourFD CloseOnExec True
+  theirFD <- Fd . fdFD <$> handleToFd theirHandle
+  return (ourHandle, CommunicationHandle theirFD)
+##else
+  trueForWinIO <-
+    return False
+##  if defined (__IO_MANAGER_WINIO__)
+      <!> return True
+##  endif
+  -- On Windows, use mkNamedPipe to create the two pipe ends.
+  alloca $ \ pfdStdInput  ->
+    alloca $ \ pfdStdOutput -> do
+      let (inheritRead, inheritWrite) = mbSwap (False, True)
+          -- If we're using WinIO, make the parent pipe end overlapped,
+          -- otherwise make both pipe ends synchronous.
+          overlappedRead  = if inheritRead  then False else trueForWinIO
+          overlappedWrite = if inheritWrite then False else trueForWinIO
+      throwErrnoIf_ (==False) "c_mkNamedPipe" $
+        -- Create one end to be un-inheritable and the other
+        -- to be inheritable, which ensures the parent end can be properly
+        -- associated with the parent process.
+        c_mkNamedPipe
+          pfdStdInput  inheritRead  overlappedRead
+          pfdStdOutput inheritWrite overlappedWrite
+      let ((ourPtr, ourMode), (theirPtr, _theirMode)) =
+            mbSwap ((pfdStdInput, ReadMode), (pfdStdOutput, WriteMode))
+      ourHANDLE  <- peek ourPtr
+      theirHANDLE <- peek theirPtr
+      -- With WinIO, we need to associate any handles we are going to use in
+      -- the current process before being able to use them.
+      return ()
+##  if defined (__IO_MANAGER_WINIO__)
+        <!> associateHandle' ourHANDLE
+##  endif
+      ourHandle <-
+##  if !defined (__IO_MANAGER_WINIO__)
+        rawFdToHandle ourMode =<< openHANDLE ourHANDLE
+##  else
+        -- NB: it's OK to call the following function even when we're not
+        -- using WinIO at runtime, so we don't use <!>.
+        rawHANDLEToHandle ourMode ourHANDLE
+##  endif
+      return $ (ourHandle, CommunicationHandle theirHANDLE)
+
+foreign import ccall "mkNamedPipe" c_mkNamedPipe ::
+    Ptr HANDLE -> Bool -> Bool -> Ptr HANDLE -> Bool -> Bool -> IO Bool
+
+##endif
+
+--------------------------------------------------------------------------------
+
+-- | A version of 'readCreateProcessWithExitCode' that communicates with the
+-- child process through a pair of 'CommunicationHandle's.
+--
+-- Example usage:
+--
+-- > readCreateProcessWithExitCodeCommunicationHandle
+-- >   (\(chTheyRead, chTheyWrite) -> proc "child-exe" [show chTheyRead, show chTheyWrite])
+-- >   (\ hWeRead -> hGetContents hWeRead)
+-- >   (\ hWeWrite -> hPut hWeWrite "xyz")
+--
+-- where @child-exe@ is a separate executable that is implemented as:
+--
+-- > main = do
+-- >   [chRead, chWrite] <- getArgs
+-- >   hRead  <- useCommunicationHandle $ read chRead
```

Comment:
`useCommunicationHandle` is not exported by the module.

Reply:
Ah yes, mentions of `useCommunicationHandle` should probably be `openCommunicationHandleRead` and `openCommunicationHandleWrite`.